xl2times/transforms.py

```diff
@@ -570,6 +570,69 @@ def process_user_constraint_table(
     return [process_user_constraint_table(t) for t in tables]
 
 
+def generate_uc_properties(
+    config: datatypes.Config,
+    tables: List[datatypes.EmbeddedXlTable],
+    model: datatypes.TimesModel,
+) -> List[datatypes.EmbeddedXlTable]:
+    """
+    Generate a dataframe containing User Constraint properties
+    """
+
+    uc_tables = [table for table in tables if table.tag == datatypes.Tag.uc_t]
+    columns = ["uc_n", "uc_desc", "region", "reg_action", "ts_action"]
+    user_constraints = pd.DataFrame(columns=columns)
+
+    for uc_table in uc_tables:
+        df = uc_table.dataframe.loc[:, ["uc_n"]].drop_duplicates(keep="first")
+        df = df.merge(
+            uc_table.dataframe.loc[:, ["uc_n", "uc_desc"]]
+            .drop_duplicates(keep="first")
+            .dropna(),
+            how="left",
+        )
+
+        for key in uc_table.uc_sets.keys():
+            if key.startswith("R_"):
+                df["reg_action"] = key
+                df["region"] = uc_table.uc_sets[key].upper().strip()
+            elif key.startswith("T_"):
+                df["ts_action"] = key
+
+        user_constraints = pd.concat([user_constraints, df], ignore_index=True)
+
+    # Use name if description is missing
+    index = user_constraints["uc_desc"].isna()
+    if any(index):
+        user_constraints["uc_desc"][index] = user_constraints["uc_n"][index]
+
+    # Exlode regions
+    # Handle allregions
+    index = user_constraints["region"].str.lower() == "allregions"
+    if any(index):
+        user_constraints["region"][index] = [model.internal_regions]
+
+    # Handle comma-separated regions
+    index = user_constraints["region"].str.contains(",").fillna(value=False)
+    if any(index):
+        user_constraints["region"][index] = user_constraints.apply(
+            lambda row: [
+                region
+                for region in str(row["region"]).split(",")
+                if region in model.internal_regions
+            ],
+            axis=1,
+        )
+
+    user_constraints = user_constraints.explode("region", ignore_index=True)
```

Comment:
We have very similar code in `process_transform_tables` and there is also a separate transform called `expand_rows` that does some of this. Can we refactor to a utility function that explodes regions that is used by all such transforms? And would it make sense to convert `allregions` into the comma-separated list of regions and then leave it like that for `expand_rows` to do the exploding for all tables?

Reply:
> And would it make sense to convert `allregions` into the comma-separated list of regions and then leave it like that for `expand_rows` to do the exploding for all tables?

Not in this case, I think, because it is used to populate `model.user_constraints` instead of being "passed down the line". If we did not explode here, we would need to explode before/on printing to GAMS. 

> We have very similar code in `process_transform_tables` and there is also a separate transform called `expand_rows` that does some of this. Can we refactor to a utility function that explodes regions that is used by all such transforms?

I am using `DataFrame.explode()` here. We have a utility function of the same name in `utils.py`, which uses  `DataFrame.explode()` internally. I guess we should rename the utility function to avoid confusion.

Reply:
All in all, I agree that we should consider how to optimise using `DataFrame.explode()`. I guess, in a separate PR?

Reply:
Actually, I believe, I will need to do it in #162 already :)

Reply:
I think the lines I wrote in `process_transform_tables` also use `DF.explode` and do very similar things:
https://github.com/etsap-TIMES/xl2times/blob/3720c7e554534131e7975753ef1b8e42c5311d97/xl2times/transforms.py#L1708-L1745
(it supports columns with region names as well, but we can enable/disable that with an argument flag in the refactored function that I am proposing)

Separate PR sounds good. But it would be good to have all the code exploding regions in one place, both for code reuse and conciseness but also so that optimizations are applied everywhere.